### PR TITLE
feat(core): add schematics to move deprecated DOCUMENT import

### PR DESCRIPTION
**Pull request by CaerusKaru**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by CaerusKaru** [comment]
After this is merged, we can finally land https://github.com/angular/angular/pull/28117

File: packages/core/schematics/test/move_document_migration_spec.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+         */
+        import {someImport} from '@angular/common';
+        import {DOCUMENT} from '@angular/platform-browser';
+      `);
+
+      runMigration();
+
+      const content = tree.readContent('/index.ts');
+
+      expect(content).toContain(`import { someImport, DOCUMENT } from '@angular/common';`);
```
Comment (devversion) [inline packages/core/schematics/test/move_document_migration_spec.ts]: I think there should be an assertion that the comment is preserved?

File: packages/core/schematics/test/move_document_migration_spec.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+      expect(content).not.toContain(`import {DOCUMENT} from '@angular/platform-browser';`);
+    });
+
+    it('should properly apply import replacement with existing import (reverse)', () => {
```
Comment (devversion) [inline packages/core/schematics/test/move_document_migration_spec.ts]: Combine this into the previous test? there can be multiple import statements for testing I think.

File: packages/core/schematics/test/google3/move_document_rule_spec.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+      import {someImport} from '@angular/common';
+    `);
+
+    runTSLint();
+
+    expectFileToContain('index.ts', `import { someImport, DOCUMENT } from '@angular/common';`);
+    expectFileNotToContain('index.ts', `import {DOCUMENT} from '@angular/platform-browser';`);
+  });
+
+  it('should properly apply import replacement with existing import (reverse)', () => {
```
Comment (devversion) [inline packages/core/schematics/test/google3/move_document_rule_spec.ts]: Should be also combined I think (same as for the non-tslint tests)

File: packages/core/schematics/migrations/move-document/google3/moveDocumentRule.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+    failures.push(new RuleFailure(
+        sourceFile, documentElement.getStart(), documentElement.getWidth(), Rule.FAILURE,
+        this.ruleName, fixes));
+
+    return failures;
+  }
+}
+
+const NEW_COMMON_TEXT = `import {DOCUMENT} from '@angular/common';`;
```
Comment (devversion) [inline packages/core/schematics/migrations/move-document/google3/moveDocumentRule.ts]: Move to top of the file?

File: packages/core/schematics/migrations/move-document/google3/moveDocumentRule.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+      `DOCUMENT is no longer exported from @angular/platform-browser in v8. Please 
+  import from @angular/common`;
+
+  applyWithProgram(sourceFile: ts.SourceFile, program: ts.Program): RuleFailure[] {
+    const visitor = new DocumentImportVisitor(program.getTypeChecker());
+    const rootSourceFiles = program.getRootFileNames().map(f => program.getSourceFile(f) !);
+    const failures: RuleFailure[] = [];
+
+    // Analyze source files by detecting imports and declarations.
+    rootSourceFiles.forEach(sourceFile => visitor.visitNode(sourceFile));
```
Comment (devversion) [inline packages/core/schematics/migrations/move-document/google3/moveDocumentRule.ts]: This should only visit the source file that is checked within `applyWithProgram`. No need to analyze all source files. We just need this for the static-query schematic as we need a global project analysis phase in order to determine query timing.

File: packages/core/schematics/migrations/move-document/google3/moveDocumentRule.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+import {DOCUMENT_TOKEN_NAME, DocumentImportVisitor} from '../document_import_visitor';
+import {addToImport, removeFromImport} from '../move-import';
+
+
+/**
+ * Rule that moves the DOCUMENT InjectionToken from the deprecation source in
+ * angular/platform-browser
+ * to the new source in angular/common. The rule also provides TSLint automatic replacements that
+ * can
+ * be applied in order to automatically migrate to the new source.
```
Comment (devversion) [inline packages/core/schematics/migrations/move-document/google3/moveDocumentRule.ts]: Comment seems to be incorrectly wrapped.

File: packages/core/schematics/migrations/move-document/document_import_visitor.ts [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+export interface Imports {
+  platformBrowserImport: ts.NamedImports|null;
+  commonImport: ts.NamedImports|null;
+  replaceText: string|null;
+  documentElement: ts.ImportSpecifier|null;
+}
+
+/**
+ * Visitor that can be used to find a set of imports in a TypeScript file.
+ */
```
Comment (devversion) [inline packages/core/schematics/migrations/move-document/document_import_visitor.ts]: Nit: Single-line comment?

File: packages/core/schematics/migrations/move-document/document_import_visitor.ts [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    // If this is not a StringLiteral it will be a grammar error
+    const moduleSpecifier = importDeclaration.moduleSpecifier as ts.StringLiteral;
+    const sourceFile = node.getSourceFile();
+    let imports = this.importsMap.get(sourceFile);
+    if (!imports) {
+      imports = {
+        platformBrowserImport: null,
+        commonImport: null,
+        replaceText: null,
+        documentElement: null,
```
Comment (devversion) [inline packages/core/schematics/migrations/move-document/document_import_visitor.ts]: What happens if there are multiple platform-browser or common imports?

File: packages/core/schematics/migrations/move-document/document_import_visitor.ts [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+      };
+    }
+
+    if (moduleSpecifier.text === PLATFORM_BROWSER_IMPORT) {
+      const documentElement = this.getDocumentElement(node);
+      if (documentElement) {
+        imports.replaceText = this.getReplaceText(documentElement);
```
Comment (devversion) [inline packages/core/schematics/migrations/move-document/document_import_visitor.ts]: Name this  `newImportSpecifierText`?

File: packages/core/schematics/migrations/move-document/BUILD.bazel [hunk 10]
```diff
@@ -0,0 +1,18 @@
+load("//tools:defaults.bzl", "ts_library")
+
+ts_library(
+    name = "move-document",
+    srcs = glob(["**/*.ts"]),
+    tsconfig = "//packages/core/schematics:tsconfig.json",
+    visibility = [
+        "//packages/core/schematics:__pkg__",
```
Comment (alexeagle) [inline packages/core/schematics/migrations/move-document/BUILD.bazel]: nit: I'm guessing you are pasting from somewhere, but this seems overspecified - just `visibility = ["//packages/core/schematics:__subpackages__"]` would work

File: packages/core/schematics/migrations/move-document/google3/BUILD.bazel [hunk 11]
```diff
@@ -0,0 +1,12 @@
+load("//tools:defaults.bzl", "ts_library")
```
Comment (alexeagle) [inline packages/core/schematics/migrations/move-document/google3/BUILD.bazel]: the migration in g3 was already complete so let's not add the google3/* files

File: packages/core/schematics/migrations/move-document/document_import_visitor.ts [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+const COMMON_IMPORT = '@angular/common';
+const PLATFORM_BROWSER_IMPORT = '@angular/platform-browser';
+export const DOCUMENT_TOKEN_NAME = 'DOCUMENT';
+
+export interface Imports {
+  platformBrowserImport: ts.NamedImports|null;
+  commonImport: ts.NamedImports|null;
+  replaceText: string|null;
+  documentElement: ts.ImportSpecifier|null;
```
Comment (devversion) [inline packages/core/schematics/migrations/move-document/document_import_visitor.ts]: Nit: Can we please add descriptions here? or even better rename this to something like `ResolvedDocumentImport`?

**Review by alexeagle — APPROVED** [review]

**Review by devversion — COMMENTED** [review]
Nice! early review. You mentioned that there is still work pending to use the `ts.Printer`. Here are some comments (mostly minor nits)

File: packages/core/schematics/migrations/move-document/BUILD.bazel (hunk 10 above)
Comment (devversion) [reply packages/core/schematics/migrations/move-document/BUILD.bazel]: This came from the other migrations. I want to limit visibility to the packages that explicitly use this. In order to enforce stricter dependencies and full dependency awareness.

File: packages/core/schematics/migrations/move-document/document_import_visitor.ts (hunk 9 above)
Comment (CaerusKaru) [reply packages/core/schematics/migrations/move-document/document_import_visitor.ts]: Ended up removing it after switching to `ts.Printer`

File: packages/core/schematics/migrations/move-document/google3/moveDocumentRule.ts (hunk 6 above)
Comment (CaerusKaru) [reply packages/core/schematics/migrations/move-document/google3/moveDocumentRule.ts]: Ended up removing as Alex suggested

File: packages/core/schematics/migrations/move-document/google3/moveDocumentRule.ts (hunk 4 above)
Comment (CaerusKaru) [reply packages/core/schematics/migrations/move-document/google3/moveDocumentRule.ts]: Ended up removing it after switching to `ts.Printer`

File: packages/core/schematics/migrations/move-document/google3/moveDocumentRule.ts (hunk 5 above)
Comment (CaerusKaru) [reply packages/core/schematics/migrations/move-document/google3/moveDocumentRule.ts]: Not needed now that we don't have the rule

**Review by alexeagle — APPROVED** [review]

**Review by devversion — APPROVED** [review]

**Review by IgorMinar — APPROVED** [review]
thanks @CaerusKaru!

**Comment by IgorMinar** [comment]
merge-assistance: global approval

**Pull request closed without merging**